File: TestP2L2.m
%% Set up model system:
clear

%Unknowns:
fdv.fitvars={'kpl' 'VIFScale'};

%Nuisance parameters; known or estimated eslewhere.
%Noise terms estimate mean value of (magnitude) noise with nonzero mean;
% set noise terms to -Inf when real or complex data is being used.
fdv.knowns={'kve','vb','T1Pyr','T1Lac','klp','Gam1','Gam2','tdel','P0','L0'};
fdv.knownvals=[0.02, 0.05, 43, 33, 0, 2.8, 4.5, 10, 0, 0];



%Describe acquisition scheme
fdv.ntp=64; % Number of timepoints
fdv.NSeg=1; % Segments per timepoint
fdv.NFlips=(fdv.ntp)*(fdv.NSeg); % Total number of excitations

%Describe temporal sampling scheme
TR=2; % constant repetition time
fdv.TR=ones(1,fdv.NFlips)*TR;
fdv.taxis=cumsum(fdv.TR)-fdv.TR(1);

%Describe excitation scheme
fdv.FlipAngle=20*ones(2,fdv.NFlips);

%Describe vascular inmput function
fdv.UseVIF=1;
fdv.VIFP=gampdf(fdv.taxis-6,2.8,4.5);
fdv.VIFL=zeros(1,fdv.NFlips);

%Placeholder for data to be fit
fdv.data=ones(2,fdv.ntp);
fdv.Name='HP Model Demo';

%Other miscellaneous
fdv.verbose=0;

%% Generate synthetic data:

parms=[0.1 1000]; %values for variables in fdv.fitvars
[EV, IV, vb]=P2L2(parms,fdv);

fdv.data=IV*vb+EV*(1-vb);

%resid=P2LBv4Err(parms,fdv);

figure(1);clf
plot(fdv.taxis,fdv.data(1,:),'g-',...
    fdv.taxis,fdv.data(2,:),'b-');
legend('Total Pyr','Total Lac')

%% Fit synthetic data and try to recover parms defined above

nruns=100;
jbopts=optimset('display','off');
LB   =[0 0];
UB   =[1 Inf];

bestresid=Inf;
bestfits = zeros(1,length(parms));
for ii=1:nruns
    %ii
    Guess=(parms*10).*rand(1,2);
    [fits,resid] = lsqnonlin(@(x) P2L2Err(x,fdv),Guess,LB,UB,jbopts);
    if resid<bestresid
        bestresid=resid;
        bestfits=fits;
    end
end

fprintf('Kpla in = %5.3f; Kpla out = %5.3f\n',parms(1),bestfits(1));

%% Plot

fh=P2L2Plot(bestfits,fdv);

File: python/TestP2L2.py
import numpy as np
import scipy.stats as stats
import scipy.optimize as opt

import P2L2
import P2L2Err
import P2L2Plot


def TestP2L2(parms=None, keys=None, values=None, knowns=None, knownVals=None):
    '''
    Script to generate synthetic data and verify that P2L2.py works as expected

    Optional inputs:
        parms                   Changes the default parameters for the calculation
        keys                    Keys for which the values of the dictionary fdv should be changed
        values                  Corresponding values to the keys of the dictionary to change
        knowns                  Names of known variables to be changed from default
        knownVals               Corresponding known variables to be changed from default

    Output:                     pyplot figure
    '''

    ### Generate synthetic data ###
    # Setting default parameters for single shot typical MDA
    #  For description of fdv keys call help(P2L2)
    fdv = {}

    fdv['fitvarNames'] = ['kpl', 'VIFScale']

    fdv['knowns'] = {
        'kve': .02,
        'vb': .05,
        'T1Pyr': 43,
        'T1Lac': 33,
        'klp': 0,
        'Gam1': 2.8,
        'Gam2': 4.5,
        'tdel': 10,
        'P0': 0,
        'L0': 0
    }
    # Check for user entered parameters
    if knowns != None:
        i = 0
        for item in knowns:
            fdv['knowns'][item] = knownVals[i]
            i+=1

    # Independent parameters
    fdv['Name'] = 'HP 2-Compartment Model Demo'
    fdv['ntp'] = 64    
    fdv['NSeg'] = 1   
    fdv['TR'] = 2  
    fdv['verbose'] = False
    # Check for user entered parameters
    if keys != None:
        i = 0
        for item in keys:
            fdv[item] = values[i]
            i+=1

    # Describe acquisition scheme 
    fdv['NFlips'] = fdv['ntp'] * fdv['NSeg'] # Total number of excitations

    # Describe temporal sampling scheme
    fdv['TR'] = np.ones( (1, fdv['NFlips']) )*fdv['TR']
    fdv['taxis'] = np.cumsum(fdv['TR']) - fdv['TR'][0]

    # Describe excitation scheme
    fdv['FlipAngle'] = 20 * np.ones( (2, fdv['NFlips']) )

    # Describe vascular input function
    #  Test data is generated synthetically from gamma distribution
    if fdv.get('UseVIF') == None:
        fdv['UseVIF'] = True
    if fdv.get('VIFP') == None:
        fdv['VIFP'] = stats.gamma.pdf(a=2.8, scale=4.5, x=fdv['taxis']-6)
    if fdv.get('VIFL') == None:
        fdv['VIFL'] = np.zeros( (1, fdv['NFlips']) )[0]


    # Placeholder data to be fit
    fdv['data'] = np.ones( (2, fdv['NFlips']) )

    # Generate synthetic data
    if parms == None:
        parms = np.array( [0.1, 1000] )
    else:
        parms = np.array( parms )
    EV, IV, vb, Mzev, Mziv = P2L2.P2L2(parms, fdv) 
    fdv['data'] = IV*vb + EV*(1-vb)

    ### Create a fit for the generated data and try to recover parameters above ###
    LB = [0, 0]
    UB = [1, np.Inf]
    guess = (parms*10) * np.random.rand(2)

    # Optimize parameters by solving least squares problem for residuals
    fit = opt.least_squares(
        fun = P2L2Err.P2L2ErrOuter(fdv),
        x0 = guess,
        bounds = (LB, UB)
    )

    popt = fit['x']
    resid = fit['fun']

    ### Plot the fits ###
    fig = P2L2Plot.plot(resid, popt, fdv)

    return fig


if __name__ == '__main__':
    TestP2L2()

